File: Actito.Loyalty.Core/Models/ActitoPass.cs
namespace ActitoSdk.Loyalty.Core.Models;

/// <summary>
/// Represents a digital pass issued by Actito.
/// </summary>
/// <remarks>
/// An <see cref="ActitoPass"/> can be used for loyalty programs, coupons, tickets, or other
/// redeemable items. It includes metadata, redemption history, and optional
/// integrations with mobile wallets like Apple Wallet or Google Pay.
/// </remarks>
public class ActitoPass
{
    /// <summary>
    /// Unique identifier of the pass.
    /// </summary>
    public string Id { get; }

    /// <summary>
    /// Optional type of the pass.
    /// </summary>
    /// <remarks>
    /// Supported pass types:
    /// 
    /// - `boarding`
    /// - `coupon`
    /// - `ticket`
    /// - `generic`
    /// - `card`
    /// </remarks>
    public string? Type { get; }

    /// <summary>
    /// Version of the pass.
    /// </summary>
    public int Version { get; }

    /// <summary>
    /// Optional Apple Wallet passbook URL or identifier.
    /// </summary>
    public string? Passbook { get; }

    /// <summary>
    /// Optional template identifier used to generate the pass.
    /// </summary>
    public string? Template { get; }

    /// <summary>
    /// Serial number of the pass.
    /// </summary>
    public string Serial { get; }

    /// <summary>
    /// Barcode value associated with the pass.
    /// </summary>
    public string Barcode { get; }

    /// <summary>
    /// Redemption behavior of the pass.
    /// </summary>
    /// <remarks>
    /// Supported redemption modes:
    /// 
    /// - `once` — the pass can be redeemed a single time
    /// - `limit` — the pass can be redeemed a limited number of times
    /// - `always` — the pass can be redeemed an unlimited number of times
    /// </remarks>
    public string Redeem { get; }

    /// <summary>
    /// History of past redemptions for this pass.
    /// </summary>
    public IList<ActitoPassRedemption> RedeemHistory { get; }

    /// <summary>
    /// Maximum number of times the pass can be redeemed.
    /// </summary>
    public int Limit { get; }

    /// <summary>
    /// Token associated with the pass for secure validation.
    /// </summary>
    public string Token { get; }

    /// <summary>
    /// Collection of key-value pairs used to add extra information to the pass.
    /// </summary>
    public IDictionary<string, object> Data { get; }

    /// <summary>
    /// Timestamp indicating when the pass was created or issued.
    /// </summary>
    public DateTime Date { get; }

    /// <summary>
    /// Optional link to save the pass to Google Pay.
    /// </summary>
    public string? GooglePaySaveLink { get; }

    /// <summary>
    /// Constructor for <see cref="ActitoPass"/>.
    /// </summary>
    public ActitoPass(string id, string? type, int version, string? passbook, string? template, string serial,
        string barcode, string redeem, IList<ActitoPassRedemption> redeemHistory, int limit, string token,
        IDictionary<string, object> data, DateTime date, string? googlePaySaveLink)
    {
        Id = id;
        Type = type;
        Version = version;
        Passbook = passbook;
        Template = template;
        Serial = serial;
        Barcode = barcode;
        Redeem = redeem;
        RedeemHistory = redeemHistory;
        Limit = limit;
        Token = token;
        Data = data;
        Date = date;
        GooglePaySaveLink = googlePaySaveLink;
    }
}

/// <summary>
/// Represents a single redemption record for an Actito pass.
/// </summary>
/// <remarks>
/// Each <see cref="ActitoPassRedemption"/> records the time and optional comments
/// when a pass was redeemed.
/// </remarks>
public class ActitoPassRedemption
{
    /// <summary>
    /// Optional comments associated with the redemption.
    /// </summary>
    public string? Comments { get; }

    /// <summary>
    /// Timestamp when the pass was redeemed.
    /// </summary>
    public DateTime Date { get; }

    /// <summary>
    /// Constructor for <see cref="ActitoPassRedemption"/>.
    /// </summary>
    public ActitoPassRedemption(string? comments, DateTime date)
    {
        Comments = comments;
        Date = date;
    }
}

File: Actito.Loyalty.iOS.Binding.Bridge/Sources/Models/ActitoPass.swift
import Foundation
import ActitoLoyaltyKit

@objc
public class ActitoPass : NSObject {
    @objc public let passId: String
    @objc public let type: String?
    @objc public let version: Int
    @objc public let passbook: String?
    @objc public let passTemplate: String?
    @objc public let serial: String
    @objc public let barcode: String
    @objc public let redeem: String
    @objc public let redeemHistory: [ActitoPassRedemption]
    @objc public let limit: Int
    @objc public let token: String
    @objc public let data: [String : Any]
    @objc public let date: Date

    @objc public init(passId: String, type: String?, version: Int, passbook: String?, passTemplate: String?, serial: String, barcode: String, redeem: String, redeemHistory: [ActitoPassRedemption], limit: Int, token: String, data: [String : Any], date: Date) {
        self.passId = passId
        self.type = type
        self.version = version
        self.passbook = passbook
        self.passTemplate = passTemplate
        self.serial = serial
        self.barcode = barcode
        self.redeem = redeem
        self.redeemHistory = redeemHistory
        self.limit = limit
        self.token = token
        self.data = data
        self.date = date
    }

    public convenience init(from pass: ActitoLoyaltyKit.ActitoPass) {
        self.init(
            passId: pass.id,
            type: pass.type?.rawValue,
            version: pass.version,
            passbook: pass.passbook,
            passTemplate: pass.template,
            serial: pass.serial,
            barcode: pass.barcode,
            redeem: pass.redeem.rawValue,
            redeemHistory: pass.redeemHistory.map { ActitoPassRedemption(from: $0) },
            limit: pass.limit,
            token: pass.token,
            data: pass.data,
            date: pass.date
        )
    }

    public func toNative() -> ActitoLoyaltyKit.ActitoPass {
        return ActitoLoyaltyKit.ActitoPass(
            id: passId,
            type: type.flatMap { ActitoLoyaltyKit.ActitoPass.PassType(rawValue: $0) },
            version: version,
            passbook: passbook,
            template: passTemplate,
            serial: serial,
            barcode: barcode,
            redeem: ActitoLoyaltyKit.ActitoPass.Redeem(rawValue: redeem)!,
            redeemHistory: redeemHistory.map { $0.toNative() },
            limit: limit,
            token: token,
            data: data,
            date: date
        )
    }
}

@objc
public class ActitoPassRedemption : NSObject {
    @objc public let comments: String?
    @objc public let date: Date

    @objc public init(comments: String?, date: Date) {
        self.comments = comments
        self.date = date
    }

    public convenience init(from redemption: ActitoLoyaltyKit.ActitoPass.Redemption) {
        self.init(
            comments: redemption.comments,
            date: redemption.date
        )
    }

    public func toNative() -> ActitoLoyaltyKit.ActitoPass.Redemption {
        return ActitoLoyaltyKit.ActitoPass.Redemption(
            comments: comments,
            date: date
        )
    }
}
